### ISSUE TITLE
Reading PMI and GD&T information from STEP files

### ISSUE THREAD

**Issue by theveloped** [tpaviot/pythonocc-core#904]
I'm wondering if it is possible to use `STEPCAFControl_Reader` to extract the dimension and tolerance information (e.g. `XCAFDoc_Dimension`) that can be embedded in a STEP file. In the examples I've found various examples that cover extracting colors but not any other information.

**Comment by tpaviot** [comment]
I guess it's possible, but never did it. If you manage to get something, please post yor feedback here

**Comment by varunkaarthik** [comment]
Hey @theveloped,

Did you manage to find a solution for this?

**Comment by onkarp12** [comment]
Hey @theveloped,
Did you manage to find a solution for this?

**Comment by tpaviot** [comment]
@onkarp12 can you please provide a test step file including PMI and GDT information for testing

**Comment by onkarp12** [comment]
[StepFilesWithPMI.zip](https://github.com/tpaviot/pythonocc-core/files/15257974/StepFilesWithPMI.zip)
@tpaviot  Please find the files. Thanks for looking into this. Appreciate it!

Regards,
Onkar

**Comment by tpaviot** [comment]
> [StepFilesWithPMI.zip](https://github.com/tpaviot/pythonocc-core/files/15257974/StepFilesWithPMI.zip) @tpaviot Please find the files. Thanks for looking into this. Appreciate it!
> 
> Regards, Onkar

:+1:

**Comment by lukasamsa** [comment]
I spent quite some time researching this topic and trying to find a solution on how to extract the graphic presentation of the GDT. Using the OCC it is pretty easy to extract the shapes of the REFERENCE shapes, which means we are left with the faces on which the GDT is appended. On the other hand it is much harder, if even possible to somehow extract the exact geometry of the GDT dimension itself.

For example:
`gdt_tool = XCAFDoc_DocumentTool.DimTolTool(doc.Main())
dim_tol_labels = TDF_LabelSequence()
gdt_tool.GetDimensionLabels(dim_tol_labels)
# gdt_tool.GetDatumLabels(labels) -> returns the Datum labels
# gdt_tool.GetGeomToleranceLabels(labels) -> returns the Geometric Tolerance labels`

extracts the labels of the dimensions, but when I tried to extract the shape of the GDT, it always returns just the reference shape. 

The .GetPresentation() function returned a TopoDS_Compound, which seems to be empty: no faces, no edges, no nothing... 

Another possible way could be this one:
![image](https://github.com/tpaviot/pythonocc-core/assets/88447657/1107f5c8-3451-4dd6-aad6-551a1249ebf3)

which I think isn't yet feasible with the pythonocc, since the NCollection_IndexedDataMap() is not wrapped... 

Similar issue was opened and discussed in this repo: [(https://github.com/usnistgov/STP2X3D/issues/7)]. 

Although this comment isn't anything new and didn't solve much, I wanted to open a discussion, which might spark an idea for someone... I think that presentation of the GDT will be an important factor, since Model Based Definitions are becoming a standard practice these days.

**Comment by tpaviot** [comment]
This place https://github.com/Open-Cascade-SAS/OCCT/blob/master/src/XDEDRAW/XDEDRAW_GDTs.cxx is certainly a good place to start with GDT. A port of these functions to python would be great

**Comment by lukasamsa** [comment]
I've seen a couple of examples using this method: 

c++:
```
Handle(TCollection_HAsciiString) aName = new TCollection_HAsciiString(argv[4]);
// Dimension
Handle(XCAFDoc_Dimension) aDimension;
if (aLabel.FindAttribute(XCAFDoc_Dimension::GetID(), aDimension))
{
  Handle(XCAFDimTolObjects_DimensionObject) anObj = aDimension->GetObject();
  anObj->SetPresentation(aPresentation, aName);
  aDimension->SetObject(anObj);
}
```
which can be easily translated to python, but I stumbled upon an issue. The `.FindAttribute(XCAFDoc_Dimension::GetID(), aDimension)` expects a handle as a second parameter. As far as I know, you can't really create a handle in python as the function expects us to, so if anyone knows any workarounds it would be great.

**Comment by mtiblow** [comment]
# Graphical PMI / GD&T from STEP AP242 in pythonocc — what works, what doesn't

This is a follow-up to [#904](https://github.com/tpaviot/pythonocc-core/issues/904).
I have a working recipe that displays graphical PMI (dimensions, GD&T,
feature control frames, datums, notes) on top of the model geometry in
the OCCT/pythonocc viewer. Reference script: `step_pmi_viewer_v2.py`.

It is not 100% complete — some files (NIST AP242 `-e3`/`-e4` tessellated
variants) hit an OCCT shaded-builder bug, and there's no clean way to
read individual `XCAFDoc_Dimension` / `XCAFDoc_GeomTolerance` semantic
attributes in this build — but it covers the common case (`-e2`-style
semantic + polyline presentation) well enough to render full GD&T
overlays from real exports out of Creo and the NIST AP242 sample set.

## Environment

| Component       | Version                                  |
| --------------- | ---------------------------------------- |
| pythonocc-core  | 7.9.0 (conda-forge)                      |
| OCCT            | 7.9.0 (bundled with the above)           |
| Python          | 3.12.7                                   |
| Qt binding      | PySide6 6.8.1 / Qt 6.8.1                 |
| OS              | Windows 11                               |
| GPU / driver    | NVIDIA RTX A2000, OpenGL 4.6, driver 546.59 |

Test files: NIST AP242 CTC/FTC test suite (e1, e2, e3 variants) plus a
handful of in-house Creo exports.

## TL;DR — the working recipe

1. Open with `STEPCAFControl_Reader` with `SetGDTMode(True)` and
   `SetViewMode(True)`. Transfer into a `TDocStd_Document("XCAF")`
   constructed with a plain Python `str`.
2. Display geometry with a single `AIS_Shape(reader.Reader().OneShape())`.
3. For every PMI label returned by `XCAFDoc_DimTolTool` (dimensions,
   geom tolerances, datums), create an `XCAFPrs_AISObject(label)` and
   display it. This gives you the constructed dim layout — extension
   lines, leaders, datum markers — but **no text** in this OCCT build.
4. For every PMI label, also walk its descendants with
   `TDF_ChildIterator(label, recursive=True)` and display every
   `ShapeTool.GetShape(child)` as `AIS_Shape`. This is where the saved
   tessellated/polyline presentation lives — text glyphs as vector
   outlines, FCF boxes, leader endpoints. Without this you see lines
   and points but no values, symbols, or letters.
5. Iterate the top-level children of `OneShape()` with `TopoDS_Iterator`
   and display anything that is **not** `TopAbs_SOLID` / `TopAbs_COMPSOLID`
   as `AIS_Shape`. This catches the datum target "+" markers and
   sketched reference geometry that ride along in the STEP compound
   but don't show up via the XCAF assembly walk.

Plus two essential bits of setup:

- Prime the OCCT font DB once at app start (see Workarounds below).
- Bump the default `Prs3d_DimensionAspect` text height — the default
  is effectively 0, so even when the recipe is correct no text renders.

## What works

Rendered cleanly on a Creo-exported plate and on the NIST AP242 CTC/FTC
**`-e2`** test files (semantic representation + polyline presentation):

- Linear / diameter / radial dimensions with `±` tolerances and
  `2X` / `4X` quantity prefixes
- Feature control frames including position, perpendicularity,
  flatness, profile, plus modifiers (`M`, `Ⓜ`, `Ⓛ`)
- Datum letter callouts in datum boxes (A–F seen)
- Datum feature symbols on faces
- Datum target "+" markers on the actual surface of the part
- Multi-line note blocks ("NOTES (UNLESS OTHERWISE SPECIFIED)" etc.)
- "2X COPLANAR SURFACES"-style annotations attached to the model

Counts on a sample NIST CTC-02 e2 file: 124 `XCAFPrs_AISObject` items
+ 124 presentation shapes from PMI children. The render takes a few
hundred milliseconds total.

## What doesn't

### 1. NIST `-e3` / `-e4` (tessellated PMI) — OCCT shaded-builder bug

OCCT 7.9.0's `AIS_Shape::Compute()` shaded-presentation builder hits
`Standard_OutOfRange: NCollection_AliasedArray::value(), out of range`
on the tessellated PMI shapes in NIST `-e3` files. The exception
propagates out of `AIS_InteractiveContext::Display`.

Per-shape fallback to wireframe mode (`AIS_Shape::SetDisplayMode(0)`)
lets the AIS object register without an exception, but the wireframe
builder also logs internal errors and the visible result is garbled
fragments rather than clean glyphs. This appears to be a genuine
OCCT bug specific to certain tessellated PMI structures — Creo struggles
with the same files visually, and it's not something a pythonocc-side
fix can solve.

### 2. `XCAFDoc_GeomTolerance` is not wrapped

```python
>>> from OCC.Core.XCAFDoc import XCAFDoc_GeomTolerance
>>> XCAFDoc_GeomTolerance()
OCC.Core.Exception.ClassNotWrappedError: XCAFDoc_GeomTolerance not wrapped
```

So the attribute-by-attribute semantic extraction path
(`label.FindAttribute(XCAFDoc_GeomTolerance.GetID(), attr)`) isn't
available for tolerances. `XCAFDoc_Dimension` and `XCAFDoc_Datum` are
wrapped, but…

### 3. `TDF_Label.FindAttribute` signature mismatch

```python
attr = XCAFDoc_Dimension()
label.FindAttribute(XCAFDoc_Dimension.GetID(), attr)
# Wrong number or type of arguments for overloaded function
# 'TDF_Label_FindAttribute'.
#   FindAttribute(Standard_GUID const&, handle<TDF_Attribute>&)
#   FindAttribute(Standard_GUID const&, Standard_Integer const, handle<TDF_Attribute>&)
```

The empty-attribute instance can't be coerced into the expected
`handle<TDF_Attribute>&` argument. So even where the attribute class IS
wrapped, you can't actually fetch the attribute through `FindAttribute`.

These two limitations together mean: in this build, you can list the
PMI labels and iterate their attached presentation shapes, but you
can't reach the structured semantic data
(`XCAFDimTolObjects_DimensionObject` etc.) to read tolerance values,
modifiers, datum references etc. programmatically. For *display*
purposes that's fine — the presentation shapes carry everything you
need to render — but it rules out semantic-to-QIF or
semantic-to-spreadsheet pipelines using this build.

### 4. `OCC.Core.Font` is not exposed

```python
>>> from OCC.Core.Font import Font_FontMgr
ModuleNotFoundError: No module named 'OCC.Core.Font'
```

The font manager and `Font_BRepFont` aren't directly available. They
are reachable indirectly via `OCC.Core.Addons.text_to_brep`, which is
enough to populate the system font DB so OCCT's text rendering finds
glyphs (see Workarounds).

### 5. Some FCF leaders don't render

On a handful of FCFs in NIST CTC-03 e2 the leader line from the FCF
box to the referenced feature didn't render in our viewer but rendered
in Creo. Walking the whole `doc.Main()` with `TDF_ChildIterator` to
catch shapes attached to labels OUTSIDE the dim/gtol/datum hierarchy
did NOT find them — it picked up coordinate-system axes and sketched
reference curves instead. The missing leaders are most likely stored
inside an `XCAFDoc_View` (saved view) and need `XCAFDoc_ViewTool` /
view-aware traversal to reach. Not chased further in this script.

## Workarounds & gotchas

### Prime OCCT's font DB

Without this, PMI text is silently skipped — extension lines and
markers render, glyphs don't. Calling `text_to_brep` once at app start
forces `Font_BRepFont` to load and registers system fonts as a side
effect.

```python
from OCC.Core.Addons import text_to_brep, Font_FontAspect_Regular
text_to_brep("x", "Arial", Font_FontAspect_Regular, 1.0, True)
```

### Set the default dimension aspect text height

XCAFPrs reads from the AIS context's default drawer when laying out
PMI text. Default height is ~0, so even with a font available nothing
renders. Bump it once.

```python
drawer = display.Context.DefaultDrawer()
drawer.TextAspect().SetHeight(5.0)
dim = drawer.DimensionAspect()
dim.MakeText3d(True)
dim.MakeTextShaded(True)
dim.TextAspect().SetHeight(5.0)
```

### Use a plain `str` for `TDocStd_Document`

```python
doc = TDocStd_Document("XCAF")     # works
```

```python
# These two patterns silently crash the process (no Python traceback,
# just an immediate exit) in this build:
doc = TDocStd_Document(TCollection_ExtendedString("XCAF"))
app = XCAFApp_Application.GetApplication()
app.NewDocument(TCollection_ExtendedString("XCAF"), doc)
```

### `qtViewer3d` auto-inits OCCT — don't double-init

`qtViewer3d.initializeGL` (fired during the first show/paint) already
calls `InitDriver` internally. Calling `InitDriver` a second time
creates a second `Viewer3d` and breaks the GL context, producing
`glBufferData ... GL_INVALID_OPERATION` errors on the first
`Display()`.

```python
if self.canvas._display is None:
    self.canvas.InitDriver()
self.display = self.canvas._display
```

### Realize the window BEFORE calling `InitDriver`

If `InitDriver` runs before the parent QMainWindow has a native window
handle, OCCT binds to a dead window and `Display()` segfaults silently.
The pattern that works:

```python
app = QApplication(sys.argv)
win = MainWindow()
win.showMaximized()
app.processEvents()      # forces native HWND creation
win.init_canvas()        # safe to InitDriver now
sys.exit(app.exec())
```

### `ShapeTool.GetShape(label)` returns `None`, not a null `TopoDS_Shape`

For labels that have no attached shape. Null-check before calling
`.IsNull()` or anything else.

```python
s = shape_tool.GetShape(label)
if s is not None and not s.IsNull():
    ...
```

### `TopoDS_Shape.HashCode()` was removed in OCCT 7.8+

So the typical `hash(shape)`-based dedup doesn't work. Use
`TopTools_IndexedMapOfShape` instead — it uses native OCCT shape
equality.

### Per-shape try / except + wireframe fallback

On `-e3` files the shaded `AIS_Shape::Compute` builder throws on a
fraction of presentation shapes. Wrap each `Display` call so a single
malformed shape doesn't abort the whole overlay, and retry in
wireframe mode (display mode 0) on failure — PMI glyphs are polyline
outlines anyway, so wireframe looks identical for them and bypasses
the broken shaded codepath.

### Silence OCCT's stderr while rendering tessellated PMI

Even when the AIS objects register OK, OCCT 7.9 prints
`Error: AIS_Shape::Compute() ... has failed` on every problematic shape.
Redirect fd 2 to `os.devnull` only around the display loop — Python
`stdout` is unaffected, so your own diagnostics still come through.

```python
from contextlib import contextmanager
import os, sys

@contextmanager
def silence_occt_stderr():
    sys.stderr.flush()
    devnull = os.open(os.devnull, os.O_WRONLY)
    saved = os.dup(2)
    os.dup2(devnull, 2)
    try:
        yield
    finally:
        sys.stderr.flush()
        os.dup2(saved, 2)
        os.close(devnull)
        os.close(saved)
```

## Three-pass display in code

```python
# Pass 1 — constructed dim geometry (no text)
for getter in (dt.GetDimensionLabels, dt.GetGeomToleranceLabels,
               dt.GetDatumLabels):
    labels = TDF_LabelSequence()
    getter(labels)
    for i in range(1, labels.Length() + 1):
        try:
            ctx.Display(XCAFPrs_AISObject(labels.Value(i)), False)
        except Exception:
            pass

# Pass 2 — saved tessellated/polyline presentation (the TEXT lives here)
for getter in (dt.GetDimensionLabels, dt.GetGeomToleranceLabels,
               dt.GetDatumLabels):
    labels = TDF_LabelSequence()
    getter(labels)
    for i in range(1, labels.Length() + 1):
        label = labels.Value(i)
        s = shape_tool.GetShape(label)
        if s is not None and not s.IsNull():
            safe_display(s)
        it = TDF_ChildIterator(label, True)
        while it.More():
            cs = shape_tool.GetShape(it.Value())
            if cs is not None and not cs.IsNull():
                safe_display(cs)
            it.Next()

# Pass 3 — datum target marks / sketched reference geometry
if one_shape.ShapeType() == TopAbs_COMPOUND:
    it = TopoDS_Iterator(one_shape)
    while it.More():
        sub = it.Value()
        if sub.ShapeType() not in (TopAbs_SOLID, TopAbs_COMPSOLID):
            safe_display(sub)
        it.Next()
```

## Reference script

The full self-contained Qt viewer (~250 lines, PySide6) is in
[`step_pmi_viewer_v2.py`](step_pmi_viewer_v2.py). It demonstrates all
of the above end-to-end:

- Realize-window-before-InitDriver pattern
- Font priming + default `Prs3d_DimensionAspect` setup
- The three-pass PMI display with per-shape fallback
- Stderr suppression scoped to the display loop

Open it, `python step_pmi_viewer_v2.py`, **File → Open STEP…**, and
point it at a NIST AP242 `-e2` test file. You should see the model
plus dim/FCF/datum/note overlays in a few hundred ms.

## Open questions

1. Is the `TDF_Label.FindAttribute` signature mismatch a binding bug
   we could fix in pythonocc-core, or is it deliberate? It would
   unlock semantic-PMI extraction.
2. Is `XCAFDoc_GeomTolerance` wrapped intentionally being left out, or
   is it just missing from the binding list?
3. Is anyone else running into the `AIS_Shape::Compute`
   `NCollection_AliasedArray` out-of-range on `-e3` files? It feels
   like an upstream OCCT issue rather than a pythonocc one.
4. Is there a documented OCCT-native path to read saved-view PMI
   presentations via `XCAFDoc_ViewTool` from Python? That would
   probably catch the missing FCF leaders.


[step_pmi_viewer_v2.py](https://github.com/user-attachments/files/28487702/step_pmi_viewer_v2.py)

**Comment by gkv311** [comment]
@mtiblow very extensive description!

I'm curious about this bit:

>Plus two essential bits of setup:
>
>-  Prime the OCCT font DB once at app start (see Workarounds below).
>-  Bump the default Prs3d_DimensionAspect text height — the default  is effectively 0, so even when the recipe is correct no text renders.
> ...
> XCAFPrs reads from the AIS context's default drawer when laying out PMI text. Default height is ~0, so even with a font available nothing renders.

But if the script displays pre-tessellated text labels, then where `Prs3d_DimensionAspect` / `Font_BRepFont` are actually used?
I can find only one place in `XCAFPrs_AISObject` that could display any text (I don't think it is related to PMI), but it is disabled by default via global flag `XCAFPrs::GetViewNameMode()`/`XCAFPrs::SetViewNameMode()`, which is not referred in your description...

> Silence OCCT's stderr while rendering tessellated PMI
> def silence_occt_stderr():

Instead of tricky manipulation with standard output streams, it is better to alter [`Message::DefaultMessenger()`](https://github.com/tpaviot/pythonocc-core/issues/1367#issuecomment-2933527442).

**Comment by mtiblow** [comment]
@gkv311

  You're right on both.

  Font / Prs3d_DimensionAspect: not used in my path. The text that renders is the pre-tessellated saved-view
  presentation (polyline/mesh geometry already in the file), shown as a plain AIS_Shape — no Font_BRepFont lookup, no
  dimension-text layout. XCAFPrs_AISObject draws only the extension/leader lines, which matches your finding that its
  only text path is XCAFPrs::SetViewNameMode() (label name, off by default, never enabled here). So those two "essential
  setup" bullets are wrong — and the premise too: a fresh Prs3d_DimensionAspect().TextAspect().Height() is 16.0 in
  7.9.0, not ~0. I dropped both and confirmed the overlay renders identically. (They'd only matter on the semantic
  PrsDim_Dimension path, which doesn't work in this build anyway.)

  stderr: agreed — switched from dup2-ing fd 2 to dropping the Message_PrinterOStream on Message::DefaultMessenger()
  around the display loop:

```python
  from contextlib import contextmanager
  from OCC.Core.Message import message, Message_PrinterOStream

  @contextmanager
  def quiet_occt():
      msgr = message.DefaultMessenger()
      seq = msgr.Printers()
      saved = [seq.Value(i) for i in range(1, seq.Size() + 1)]
      msgr.RemovePrinters(Message_PrinterOStream.get_type_descriptor())
      try:
          yield
      finally:
          for p in saved:
              msgr.AddPrinter(p)
```

  Notes and script updated. Thanks for the careful read.

[step_pmi_viewer_v2.py](https://github.com/user-attachments/files/28528084/step_pmi_viewer_v2.py)
